server/protocols/webadmin/webadmin-data-jmap/src/main/java/org/apache/james/webadmin/data/jmap/MessagePreviewCorrector.java

```diff
@@ -0,0 +1,166 @@
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.MessageManager;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.FetchGroup;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.mailbox.model.MessageRange;
+import org.apache.james.mailbox.model.MessageResult;
+import org.apache.james.mailbox.model.search.MailboxQuery;
+import org.apache.james.mime4j.dom.Message;
+import org.apache.james.mime4j.stream.MimeConfig;
+import org.apache.james.user.api.UsersRepository;
+import org.apache.james.user.api.UsersRepositoryException;
+import org.apache.james.util.html.HtmlTextExtractor;
+import org.apache.james.util.mime.MessageContentExtractor;
+import org.apache.james.util.streams.Iterators;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.Throwing;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class MessagePreviewCorrector {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MessagePreviewCorrector.class);
+
+    static class Context {
+        private final AtomicLong processedUserCount;
+        private final AtomicLong processedMessageCount;
+        private final AtomicLong failedUserCount;
+        private final AtomicLong failedMessageCount;
+
+        Context() {
+            failedUserCount = new AtomicLong();
+            processedMessageCount = new AtomicLong();
+            processedUserCount = new AtomicLong();
+            failedMessageCount = new AtomicLong();
+        }
+
+        long getProcessedUserCount() {
+            return processedUserCount.get();
+        }
+
+        long getProcessedMessageCount() {
+            return processedMessageCount.get();
+        }
+
+        long getFailedUserCount() {
+            return failedUserCount.get();
+        }
+
+        long getFailedMessageCount() {
+            return failedMessageCount.get();
+        }
+
+        boolean noFailure() {
+            return failedMessageCount.get() == 0 && failedUserCount.get() == 0;
+        }
+    }
+
+    private final UsersRepository usersRepository;
+    private final MailboxManager mailboxManager;
+    private final MessagePreviewStore messagePreviewStore;
+    private final MessageContentExtractor messageContentExtractor;
+    private final HtmlTextExtractor htmlTextExtractor;
+
+    @Inject
+    MessagePreviewCorrector(UsersRepository usersRepository, MailboxManager mailboxManager, MessagePreviewStore messagePreviewStore, MessageContentExtractor messageContentExtractor, HtmlTextExtractor htmlTextExtractor) {
+        this.usersRepository = usersRepository;
+        this.mailboxManager = mailboxManager;
+        this.messagePreviewStore = messagePreviewStore;
+        this.messageContentExtractor = messageContentExtractor;
+        this.htmlTextExtractor = htmlTextExtractor;
+    }
+
+    Mono<Void> correctAllPreviews(Context context) {
```

Comment:
having 3 methods `correctAllPreviews` doesn't help when reading.
Could we have a top-level method with a few map/flatMap/concatMap steps that make it obvious what the needed steps are and delegate to well-named methods for each step?